src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -1226,40 +1268,41 @@ completeMerge ::
   -> m ()
 completeMerge mergeVar mergeKnownCompletedVar = do
     modifyMVarMasked_ mergeVar $ \case
-      mr@CompletedMerge{} -> pure $! mr
+      mrs@CompletedMerge{} -> pure $! mrs
       (OngoingMerge rs _spentCreditsVar m) -> do
-        -- TODO: we'll likely move away from this style of reference counting,
-        -- so this code will change in the future.
-        RefCount n <- Merge.readRefCount m
-        let !n' = fromIntegralChecked n
-        V.forM_ rs $ \r -> Run.removeReferenceN r n'
+        -- first try to complete the merge, in case it fails
         r <- Merge.complete m
-        Merge.removeReferenceN m n'
+        V.forM_ rs Run.removeReference
         -- Cache the knowledge that we completed the merge
         writeMutVar mergeKnownCompletedVar MergeKnownCompleted
         pure $! CompletedMerge r
 
-{-# SPECIALISE expectCompletedMerge :: MergingRun IO h -> IO (Run IO h) #-}
+{-# SPECIALISE expectCompletedMerge :: IncomingRun IO h -> IO (Run IO h) #-}
 expectCompletedMerge ::
      (MonadMVar m, MonadSTM m, MonadST m, MonadMask m, MonadFix m)
-  => MergingRun m h -> m (Run m h)
-expectCompletedMerge (SingleRun r) = pure r
-expectCompletedMerge (MergingRun _ _ numInputEntries _ totalStepsVar@(TotalStepsVar var) mergeKnownCompletedVar mergeVar) = do
-    mergeKnownCompleted <- readMutVar mergeKnownCompletedVar
+  => IncomingRun m h -> m (Run m h)
+expectCompletedMerge (Single r) = pure r
+expectCompletedMerge (Merging (mr@MergingRun {..})) = do
+    knownCompleted <- readMutVar mergeKnownCompleted
     -- The merge is not guaranteed to be complete.
-    when (mergeKnownCompleted == MergeMaybeCompleted) $ do
-      totalSteps <- readPrimVar var
-      isMergeDone <- stepMerge mergeVar totalStepsVar (Credit (unNumEntries numInputEntries - totalSteps))
-      when isMergeDone $ completeMerge mergeVar mergeKnownCompletedVar
+    when (knownCompleted == MergeMaybeCompleted) $ do
+      totalSteps <- readPrimVar (getTotalStepsVar mergeStepsPerformed)
+      isMergeDone <- stepMerge mergeState mergeStepsPerformed (Credit (unNumEntries mergeNumEntries - totalSteps))
+      when isMergeDone $ completeMerge mergeState mergeKnownCompleted
       -- TODO: can we think of a check to see if we did not do too much work here?
-    withMVar mergeVar $ \case
+    r <- withMVar mergeState $ \case
       CompletedMerge r -> pure r
       OngoingMerge{} -> do
-        -- If the algorithm /still/ finds an ongoing merge here, then it is a bug in
+        -- If the algorithm finds an ongoing merge here, then it is a bug in
         -- our merge sceduling algorithm. As such, we throw a pure error.
         error "expectCompletedMerge: expected a completed merge, but found an ongoing merge"
+    -- return a fresh reference to the run
+    -- TODO TempRegistry?
```

Comment:
Empty newline

Reply:
Ah, maybe I still want to address this TODO in this PR.